### PR TITLE
Fix concurrent execution under Yath

### PR DESCRIPTION
**Pull request by nohuhu**

### CODE REVIEW

File: lib/Test/PostgreSQL.pm [hunk 1]
```diff
@@ -116,7 +116,42 @@ has socket_dir => (
...
+        # TODO Come up with something better if this ever cause
+        # any trouble. Straightforward approach with using File::Temp
+        # that consults $TMPDIR environment variable is exactly
+        # the cause for the Unix socket path being too long;
+        # other approaches at sniffing what system we're running on
+        # and what temporary paths are available also fail at various
+        # edge cases, which is exactly the reason File::Temp was created
+        # in the first place. So, no easy answer here; simply falling back
+        # to the timeless default.
+        DIR => '/tmp',
```
Comment (nohuhu) [inline lib/Test/PostgreSQL.pm:147]: I made this change assuming that we only support Unixy systems; only after pushing the branch I realized that we do try to support Windows. @SysPete Are we actually using Unix sockets on Windows? It's been a long time since I've done any development on that platform so I'm not sure if Unix sockets are supported there. If they are not then `socket_path` is likely irrelevant and this hardcoded value shouldn't cause any trouble; otherwise I'll need to come up with something more compatible because breaking existing Windows support is not my goal.

File: lib/Test/PostgreSQL.pm [hunk 2]
```diff
@@ -335,6 +370,7 @@ method DEMOLISH($in_global_destruction) {
     if (defined $self->pid && $self->_owner_pid == $$) {
       $self->stop
     }
+    undef $self->{_socket_dir};
```
Comment (nohuhu) [inline lib/Test/PostgreSQL.pm:373]: This _shouldn't_ be necessary, I just felt better about cleaning it up explicitly since I'm not very familiar with object destruction sequence of the library we're using. It's been a long time since I've done any development on it.

File: lib/Test/PostgreSQL.pm [hunk 3]
```diff
@@ -792,7 +844,7 @@ conflicting TCP ports on localhost.]
 
 =head2 socket_dir
 
-Unix socket directory to use if L</unix_socket> is true. Default is C<$basedir/tmp>.
+Unix socket directory to use if L</unix_socket> is true. Default is C<$base_dir/tmp>.
```
Comment (nohuhu) [inline lib/Test/PostgreSQL.pm:847]: This is a drive-by fix.

File: lib/Test/PostgreSQL.pm [hunk 4]
```diff
@@ -825,7 +877,7 @@ Extra args to be appended to L</initdb_args>. Default is empty.
 
 =head2 pg_config
 
-Configuration to place in C<$basedir/data/postgresql.conf>. Use this to override
+Configuration to place in C<$base_dir/data/postgresql.conf>. Use this to override
```
Comment (nohuhu) [inline lib/Test/PostgreSQL.pm:880]: Ditto.

File: lib/Test/PostgreSQL.pm [hunk 5]
```diff
@@ -389,16 +425,26 @@ method start() {
...
-  my $port = $self->base_port;
+
   srand(); # Re-seed the RNG in case the caller forked the process
+
+  # There is a significant chance that there might be more than
+  # one Test::Postgresql test in flight so we better randomize
+  # the port before even trying to start the first time. Does
+  # no harm if there is no concurrent Postgres running; if there is
+  # then our process will start up a bit faster.
+  my $port = $self->base_port + int(rand(10)) + 1;
```
Comment (nohuhu) [inline lib/Test/PostgreSQL.pm:436]: This might muddy the matters a bit since this change is also related to the issue; that said, I'm pretty confident that the real cause for the problem was in Unix socket path being too long since that was what Postgres was complaining about in its log.

**Comment by nohuhu** [comment]
@TJC @SysPete Any concerns/suggestions/feedback? Looks like this fix does resolve https://github.com/TJC/Test-postgresql/issues/40, and I can surmise that it resolves https://github.com/TJC/Test-postgresql/issues/37 as well since that's basically the same issue.

**Comment by TJC** [comment]
Sorry, I've been meaning to check the PR out, but I was having a break from Friday->Monday. Will check it out shortly.

File: lib/Test/PostgreSQL.pm [hunk 6]
```diff
@@ -116,7 +116,42 @@ has socket_dir => (
   is => "ro",
   isa => Str,
   lazy => 1,
-  default => method () { File::Spec->catdir( $self->base_dir, 'tmp' ) },
+  default => method () {
+    my $dir = File::Spec->catdir( $self->base_dir, 'tmp' );
```
Comment (TJC) [inline lib/Test/PostgreSQL.pm:120]: This anonymous default method has ended up quite long now -- if it's easy, could it get refactored out to a named method and just referenced from here? It might help readability.

File: lib/Test/PostgreSQL.pm [hunk 7]
```diff
@@ -116,7 +116,42 @@ has socket_dir => (
...
+
+    # This magic number is based on the Unixy systems limit for
+    # Unix socket path lengths: on Linuxes it is at 108 characters,
+    # MacOS allows no more than 103. Considering that we are
```
Comment (TJC) [inline lib/Test/PostgreSQL.pm:124]: Really appreciate the comments like this!

File: lib/Test/PostgreSQL.pm (hunk 5 above)
Comment (TJC) [reply lib/Test/PostgreSQL.pm:436]: Seems like a good plan though.

File: lib/Test/PostgreSQL.pm [hunk 8]
```diff
@@ -389,16 +425,26 @@ method start() {
...
+  # no harm if there is no concurrent Postgres running; if there is
+  # then our process will start up a bit faster.
+  my $port = $self->base_port + int(rand(10)) + 1;
+
   # try by incrementing port number until PostgreSQL starts
   while (1) {
-    my $good = try {
+    my $good;
+    try {
+      #warn "Trying to start postgres on port $port...";
```
Comment (TJC) [inline lib/Test/PostgreSQL.pm:442]: Rather than comment out the log messages, maybe set them to only print if a debug flag is set?

File: lib/Test/PostgreSQL.pm [hunk 9]
```diff
@@ -425,12 +471,18 @@ method _try_start($port) {
...
 
         my $pid_path = File::Spec->catfile( $self->base_dir, 'data', 'postmaster.pid' );
 
-        open( my $pidfh, '<', $pid_path )
-          or die "Failed to open $pid_path: $!";
+        my $pidfh;
+        if (not open($pidfh, '<', $pid_path)) {
+            #open( my $logh, '<', $logfile );
+            #my @loglines = <$logh>;
+            #warn "Postgres log: \n" . join("\n", @loglines);
```
Comment (TJC) [inline lib/Test/PostgreSQL.pm:483]: Similarly, maybe see if there's a neater way to enable/disable debug output?

**Review by TJC — APPROVED** [review]
Looks pretty good -- happy for you to merge this, but I also left some comments for a couple of potential little improvements

**Comment by SysPete** [comment]
Apologies for my silence: I started on a new client project last Thursday and have been swamped. Since @TJC has approved then I have no objections, but I will followup in the next few days with my thoughts.

**Comment by grahamc** [comment]
Hey @SysPete have you had a few minutes to take another look? Thanks!

**Comment by grahamc** [comment]
I ended up tagging this as release-1.28-1 on my fork: https://github.com/grahamc/Test-postgresql/releases/tag/release-1.28-1 and updated my project to fetch it there instead of from cpan. Thank you @nohuhu for the patch, and everyone else for the library!

**Comment by TJC** [comment]
Hey folks, looks like this this fell through the cracks - we should probably get around to merging and releasing if it looks good.

**Comment by TJC** [comment]
Although I do note that we still have some unresolved comments on the PR that need to get addressed first. (eg. Windows support; plus some minor bits)

**Comment by nohuhu** [comment]
@TJC Yep, Windows compat basically is the reason I haven't pushed this to a release yet. @grahamc Apologies for inconvenience, I'd rather have @SysPete's explicit approval before rolling this out to every user.

**Comment by SysPete** [comment]
Apologies (yet again) but I will definitely have a look in the morning.

File: lib/Test/PostgreSQL.pm (hunk 8 above)
Comment (SysPete) [reply lib/Test/PostgreSQL.pm:442]: I agree here, although I don't want to see these warnings by default, a debug flag which causes them to be printed to STDOUT would be a nice improvement.

**Review by SysPete — APPROVED** [review]
Good to merge, with comments that mostly just agree with @TJC :grinning: Thank you @nohuhu!

File: lib/Test/PostgreSQL.pm (hunk 6 above)
Comment (nohuhu) [reply lib/Test/PostgreSQL.pm:120]: Yeah I was thinking about that when I was working on this PR... In the end decided against splitting it out since the length of the method is basically all in the comments, there's not that much code in it. If we split it out, that's going to be just another long function, although a named one.

File: lib/Test/PostgreSQL.pm (hunk 8 above)
Comment (nohuhu) [reply lib/Test/PostgreSQL.pm:442]: There are pros and cons to this... It's been some time since I worked on this and if I remember correctly, I did consider adding a flag and decided not to, because these warnings are needed _very_ rarely and only by us package developers while troubleshooting; it shouldn't be much of a burden to uncomment a line of code you're working with.

On the other hand, a flag would need to be checked at run time and that's a cost that every user out there will have to pay constantly with very low probability of ever needing to turn it on. A small cost for sure but these things can add up to seconds in large test runs.

**Comment by nohuhu** [comment]
Ok everybody seems to be good with this.
